pkg/reconciler/config/store.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"reflect"
+	"sync/atomic"
+
+	"github.com/knative/pkg/configmap"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type Logger interface {
+	Infof(string, ...interface{})
+	Fatalf(string, ...interface{})
+	Errorf(string, ...interface{})
+}
+
+type UntypedStore struct {
+	name   string
+	logger Logger
+
+	storages     map[string]*atomic.Value
+	constructors map[string]reflect.Value
+}
+
+func NewUntypedStore(
+	name string,
+	logger Logger,
+	constructors ...interface{}) *UntypedStore {
```

Comment:
> The reflection here makes me a bit uncomfortable

It's solely needed for the constructor methods - not the keys

> The map enforces the argument pairing, and can be called with only slightly more code:

Yeah I was optimizing for succinctness. Plus being comfortable with the non-idiomatic Go since it's pattern in Objective-C [example](https://developer.apple.com/documentation/foundation/nsdictionary/1574181-dictionarywithobjectsandkeys)

I'll switch to the map if that makes it clearer.

> I suspect at least some of the motivation for reflection is that Go doesn't allow

Yup exactly - the typing is in go is very strict.

https://developer.apple.com/documentation/foundation/nsdictionary/1574181-dictionarywithobjectsandkeys